### PR TITLE
Optimize Read Channels

### PR DESCRIPTION
**Pull request by Maithem**
Use a single channel per segment for uncached reads.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ba6a205**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #940 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/w3c5VqEwFKz1mhwVmLOJ2RoTOzbpAfH3)

**Comment by Maithem** [comment]
@no2chem Rebased.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public void close() {
+        Set<FileChannel> channels =
+                new HashSet(Arrays.asList(writeChannel, readChannel, trimmedChannel, pendingTrimChannel));
+        for (FileChannel channel : channels) {
+            try {
+                channel.force(true);
+                channel.close();
+                channel = null;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: MAJOR Remove this useless assignment to local variable "channel". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1854)

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    @NonNull
+    final FileChannel trimmedChannel;
+
+    @NonNull
+    final FileChannel pendingTrimChannel;
+
+    @NonNull
+    String fileName;
+
+    private Map<Long, AddressMetaData> knownAddresses = new ConcurrentHashMap();
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: MINOR Remove the "knownAddresses" field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1450)

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    private Set<Long> trimmedAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: MINOR Remove the "trimmedAddresses" field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1450)

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    private Set<Long> pendingTrims = Collections.newSetFromMap(new ConcurrentHashMap<>());
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: MINOR Remove the "pendingTrims" field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1450)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [StreamLogFiles.java#L245](https://github.com/CorfuDB/CorfuDB/blob/5db09d786cfccc2044f83d47328c5702d5abbbd7/infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java#L245): Remove this unnecessary cast to "Collection". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1905)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **5db09d7**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #940 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/NTguSASSgPyfXuwunUPA0QbfUbGb1xxC)

**Pull request merged**